Iter19

kuzminal opened the pull request:
Инкремент 18 и 19
Добавил линтер, добавил вывод флагов сборки

Commit:
Добавил свой анализатор, который ищет функцию main и вызов в ней Exit

Commit:
Вернул случайно удаленные из коммита тесты

Commit:
Добавил вывод флагов сборки
Добавил покрытие тестами

Commit:
Поправил предупреждения линтера

Commit:
Поправил теств базы, добавил testcontainers

Commit:
Поправил теств базы, добавил testcontainers

Commit:
Добавил логгер для автотестов

Commit:
Добавил логгер для автотестов

Commit:
Добавил логгер для автотестов

Commit:
Добавил логгер для автотестов

Commit:
Добавил логгер для автотестов

Commit:
Добавил логгер для автотестов

Commit:
Добавил логгер для автотестов

Commit:
Вернул документацию, утерянную при коммите

Commit:
Вернул документацию, утерянную при коммите

Commit:
ПОправил статикчеки

Commit:
ПОправил статикчеки

avhitrov commented on cmd/shortener/main.go:
```diff
@@ -4,18 +4,47 @@ import (
+	"net/http"
+	"os"
+	"text/template"
+	"time"
```
Предлагаю на этом этапе причесать импорты. Гайдлайн: https://t.me/bushigo/27

avhitrov commented on cmd/shortener/main.go:
```diff
@@ -31,7 +60,9 @@ func run() error {
-	defer storage.Close()
+	defer func(storage store.AuthStore) {
+		_ = storage.Close()
+	}(storage)
```
Это мы так линтер пытаемся умаслить? Не понимаю смысла замены.

avhitrov commented on cmd/shortener/main.go:
```diff
@@ -40,16 +71,18 @@ func run() error {
 	if config.DatabaseDSN != "" {
-		rdb, err := newRDBStore(ctx, config.DatabaseDSN)
-		if err != nil {
+		logrus.Printf("Create DB storage %s", config.DatabaseDSN)
```
Не в качестве претензии, но для расширения кругозора. Такой вывод в продуктовом коде может быть оформлен только в виде logrus.Debug. В общем, на уровне отладки, которую в проде отключают (остается уровень info или выше). Кроме того отправка в консоль DatabaseDSN, который очень часто в себе несет пароль, равнозначна рисованию пароля на бумажке, прикрепленной к монитору. Очень не поощряется ИБ.

avhitrov commented on cmd/staticlint/multichecker.go:
```diff
@@ -0,0 +1,84 @@
+	"honnef.co/go/tools/staticcheck"
+)
+
+// OsExitCheckAnalyzer проверка на наличие os.Exit в футнкции main в пакете main
```
Пробегись глазами по всем комментариям на предмет опечаток. Очень огорчает видеть небрежно украшенный код.

avhitrov commented on internal/store/memory.go:
```diff
@@ -110,21 +133,26 @@ func (m *InMemory) LoadUsers(_ context.Context, uid uuid.UUID) (urls map[string]
+	userID := uid.String()
 	for _, id := range ids {
-		userID := uid.String()
+		m.mu.Lock()
```
Lock-Unlock - довольно "тяжелые" вычислительные операции. Без крайней необходимости не следует их запихивать внутрь цикла. Эффективней заблокировать до цикла и освободить лок после.

avhitrov commented on internal/store/file.go:
```diff
@@ -56,12 +57,14 @@ func NewFileStore(filepath string) (*FileStore, error) {
+// Save сохраняем ссылку в файловом хранилище.
 func (f *FileStore) Save(_ context.Context, u *url.URL) (id string, err error) {
 	id = fmt.Sprintf("%x", len(f.store.Hot))
 	f.store.Hot[id] = u
```
В in-memory, значит, блокировки расставлены... А почему здесь мапы не защищены от конкурентного доступа? Насколько я понимаю, файловый сторадж реализует тот же интерфейс и, значит, будет работать в такой же агрессивной конкурентной среде. И рано или поздно словит DATA RACE и уронит приложение.

avhitrov reviewed (changes requested):
1. Давай причешем импорты
2. Пробежаться по комментариям на предмет опечаток
3. Гонки в файловом сторадже
4. Прочие замечания по коду

kuzminal committed:
Поправил замечания:
- отформатировал импорт;
- добавил мьютексы в файловое хранилище;
- устранил некоторые опечатки;
- убрал DSN в логах

avhitrov reviewed (approved):
Отлично!
